packages/front-end/components/Experiment/ResultsDownloadButton.tsx

```diff
@@ -0,0 +1,131 @@
+import {
+  ExperimentReportResultDimension,
+  ExperimentReportVariation,
+} from "back-end/types/report";
+import React, { useMemo } from "react";
+import { FaFileExport } from "react-icons/fa";
+import { useDefinitions } from "../../services/DefinitionsContext";
+import { ExperimentTableRow, getRisk } from "../../services/experiments";
+import { Parser } from "json2csv";
+import { MetricInterface, MetricStats } from "back-end/types/metric";
+
+type CsvRow = {
+  date?: string;
+  dimension?: string;
+  metricName?: string;
+  variantName?: string;
+  riskOfChoosing?: number;
+  users?: number;
+  count?: number;
+  conversionRate?: number;
+  chanceToBeatControl?: number | null;
+  percentChange?: number | null;
+};
+
+type Variation = {
+  buckets?: { x: number; y: number }[];
+  ci?: [number, number];
+  cr: number;
+  risk?: [number, number];
+  stats?: MetricStats;
+  users: number;
+  value: number;
+};
+
+type Row = {
+  label?: string;
+  metric?: MetricInterface;
+  rowClass?: string;
+  variations?: Variation[];
+};
+
+export default function ResultsDownloadButton({
+  results,
+  metrics,
+  variations,
+  trackingKey,
+  dimension,
+}: {
+  results?: ExperimentReportResultDimension[];
+  metrics: string[];
+  variations: ExperimentReportVariation[];
+  trackingKey: string;
+  dimension: string;
+}) {
+  const { getMetricById, getDimensionById, ready } = useDefinitions();
+
+  const json2csvParser = new Parser();
+
+  const csvRows: CsvRow[] = [];
+
+  const dimensionName = getDimensionById(dimension);
+
+  if (dimension === "pre:date") {
+    results.sort((a, b) => (a.name > b.name ? 1 : b.name > a.name ? -1 : 0));
+  }
```

Comment:
`sort` mutates the original array.  You typically don't want a component to mutate it's array/object props since the parent components are usually not expecting that.

For example:

```tsx
const items = ["b", "a"];
return (
  {showChild && <ChildComponent items={items}/>}
  {items.join(" ")}
)
```

If `<ChildComponent>` called `items.sort`, then this would sometimes render `a b` and sometimes `b a` depending on whether or not `showChild` is true, which is weird and unexpected.

A better option is to create a copy of the array first.  So:
```ts
results = [...results];
results.sort(
```

Reply:
Mmmhmm that makes a lot of sense. I've refactored so it now sorts the rows just before converting it to a csv, instead of sorting/mutating the `results`.